src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs

```diff
@@ -0,0 +1,183 @@
+            CancellationToken cancellationToken)
+        {
+            Debug.Assert(clientCapabilities.HasVisualStudioLspCapability());
+
+            // The VS LSP client supports streaming using IProgress<T> on various requests.
+            // However, this is not yet supported through Live Share, so deserialization fails on the IProgress<T> property.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1043376 tracks Live Share support for this (committed for 16.6).
+            var document = solution.GetDocumentFromURI(referenceParams.TextDocument.Uri);
+            if (document == null)
+            {
+                return new LSP.VSReferenceItem[] { };
+            }
+
+            var findUsagesService = document.GetRequiredLanguageService<IFindUsagesLSPService>();
+            var position = await document.GetPositionFromLinePositionAsync(
+                ProtocolConversions.PositionToLinePosition(referenceParams.Position), cancellationToken).ConfigureAwait(false);
+            var context = new SimpleFindUsagesContext(cancellationToken);
+
+            // Finds the references for the symbol at the specific position in the document, pushing the results to the context instance.
+            await findUsagesService.FindReferencesAsync(document, position, context).ConfigureAwait(false);
+
+            return await GetReferenceItemsAsync(referenceParams, context, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<LSP.VSReferenceItem[]> GetReferenceItemsAsync(
+            LSP.ReferenceParams request,
+            SimpleFindUsagesContext context,
+            CancellationToken cancellationToken)
+        {
+            // Mapping each reference to its definition
+            var definitionMap = new Dictionary<DefinitionItem, List<SourceReferenceItem>>();
+            foreach (var reference in context.GetReferences())
+            {
+                if (!definitionMap.ContainsKey(reference.Definition))
+                {
+                    definitionMap.Add(reference.Definition, new List<SourceReferenceItem>());
+                }
+
+                definitionMap[reference.Definition].Add(reference);
+            }
+
+            // NOTE: Parts of FAR currently do not display correctly due to a bug in LSP.VSReferenceItem.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1088938/
+            using var _ = ArrayBuilder<LSP.VSReferenceItem>.GetInstance(out var referenceItems);
+
+            // Each reference is assigned its own unique id
+            var id = 0;
+            foreach (var (definition, references) in definitionMap)
+            {
+                // Creating the reference item that corresponds to the definition
+                var definitionItem = await GenerateReferenceItem(
+                    id, definitionId: id, definition.SourceSpans.FirstOrDefault(), context, definition.DisplayableProperties,
```

Comment:
small problem.  what if the definition is in metadata and has no source-spans.  does this work properly?  can you manually test that a metadata-find-refs still has proper results that can be grouped/navigated.  Also, ensure we don't crash if the user tries to navigate to the definition itself

Reply:
This was a great catch, the code didn't do this properly before, but I believe it should be fixed now. Thanks!